rfd/0080-piv-private-key.md

```diff
@@ -0,0 +1,316 @@
+	TouchRequired bool `json:"touch_required"`
+  // json blob containing the resulting attestation object. 
+  //
+  // For yubikey, this will look like https://pkg.go.dev/github.com/go-piv/piv-go@v1.9.0/piv#Attestation
+  // In the future, we can check the attesation object for additional information, like PIN policy,
+  // model type, serial number, etc.
+  AttestationObject []byte `json:"attestation"`
+}
+```
+
+### Certificate private key policy extension
+
+When the Auth Server receives a new certificate request, it will check the cluster auth preference and the user's role options to get their required private key policy. If their policy is `hardware_key`, then the Auth Server will look for a valid attestation in the backend associated with the given public key. If their policy is `hardware_key_touch`, then the Auth Server will also check that the attestation has `TouchRequired: true`.
+
+If the attestation checks are successful, then the Auth Server will sign the TLS and SSH certificates with the new `private_key_policy` TLS extra subject name and Extension respectively. The Auth Server's Authorizer, which is used for all API requests, will check certificates for this extension before authorizing requests.
+
+### Client changes
+
+Teleport clients will need the ability to connect to a user's hardware key, generate/retrieve private keys, and use those keys for cryptographical operations. 
+
+#### Private key policy discovery
+
+Teleport clients should be able to automatically determine if a user requires a hardware private key for login to avoid additional UX concerns. For this, we will introduce the `GetPrivateKeyPolicy` rpc.
+
+```proto
+service AuthService {
+  rpc GetPrivateKeyPolicy(GetPrivateKeyPolicyRequest) returns (GetPrivateKeyPolicyResponse);
+}
+
+// The user will be pulled from the request certificate, so request message can be empty for now.
+message GetPrivateKeyPolicyRequest {}
+
+message GetPrivateKeyPolicyResponse {
+    PrivateKeyPolicy policy = 1;
+}
+
+enum PrivateKeyPolicy {
+    PRIVATE_KEY_POLICY_NONE = 0;
+    PRIVATE_KEY_POLICY_HARDWARE_KEY = 1;
+    PRIVATE_KEY_POLICY_HARDWARE_KEY_TOUCH = 2;
+}
+```
+
+In order for a Teleport Client to call `GetPrivateKeyPolicy`, it first needs a set of valid certificates. Unlike `IsMFARequired`, which issues an MFA certificate Reissue request after logging in with non-MFA certificates, we need to call `GetPrivateKeyPolicy` on the initial login.
+
+One option is to make add an endpoint in the Proxy Web API which can be called before login. In order to authenticate the user, we would need them to login with their login credentials. For MFA and Password login, we can potentially reuse these password credentials to intiate the actual login directly after retrieving the user's private key policy requirement. However, for sso login and some edge cases is MFA login (OTP token timing, between requests, etc.), the user would be required to perform the login twice.
```

Comment:
> we can potentially reuse these password credentials

I'd be sad if replaying like that worked, but I expect for all cases except pure-password login it won't.